### PR TITLE
feat(world): add system call event

### PR DESCRIPTION
**Pull request by alvrs**
Draft for #1344

### CODE REVIEW

File: packages/world/gas-report.json [hunk 1]
```diff
@@ -231,67 +231,73 @@
...
     "test": "testCallFromTimeboundDelegation",
     "name": "call a system via a timebound delegation",
-    "gasUsed": 34791
+    "gasUsed": 36952
+  },
+  {
+    "file": "test/SystemCall.t.sol",
+    "test": "testGasSystemCallEvent",
+    "name": "emit a system call event",
+    "gasUsed": 3158
```
Comment (alvrs) [inline packages/world/gas-report.json:258]: Unfortunately it seems like emitting this event is pretty significant. With the current design we'd have to emit it twice per system call.

**Comment by ludns** [comment]
one thing we could do is not using events at all (which cost gas) and instead use traces as a way to get these "events" into the sync stack.
traces allow you to see EVM CALL and DELEGATECALL with their calldata, which is enough to reconstruct these events.

File: packages/world/gas-report.json (hunk 1 above)
Comment (ludns) [reply packages/world/gas-report.json:258]: why is it so expensive? it seems like the data is not that big

**Comment by alvrs** [comment]
It seems like emitting an event with sufficient data to connect it to a specific system function call (`bytes32 systemId, bytes4 functionSelector`) would still cost 1440 gas per event, and we'd need two events (start and end) to be able to assign every store event to a system call (for situations where a system updates a table, then calls another system, and then updates a table again). 

2880 gas overhead for every system call for this feature is definitely too much. One alternative approach would be to not emit the events for every system call, but only for systems where it is needed. The `Systems` table still has space in the first storage slot of static data, so we could add a `emitSystemCallEvents` flag and allow it to be set when registering systems.

**Comment by alvrs** [comment]
> one thing we could do is not using events at all (which cost gas) and instead use traces as a way to get these "events" into the sync stack.
traces allow you to see EVM CALL and DELEGATECALL with their calldata, which is enough to reconstruct these events.

Yeah thought about this as well. Unfortunately I don't think traces are supported by all RPC nodes and I assume they're pretty expensive operations, but we could this feature to the indexer (so at least it's one trace per system call, not one trace per system call per client). After seeing the gas cost of these events I do prefer a solution via traces.

**Comment by holic** [comment]
Given that folks can manually backfill for previous usages of system call stream via offchain tables or (soon) custom events, I think it's fine punting on this and being more of a feature you can enable if your RPC supports tracing.

**Comment by ludns** [comment]
it's interesting that custom events will have the same "significant increased gas usage" issue; i wonder if that is going to tip the balance back to system calls via traces again.

**Comment by alvrs** [comment]
Seems like we're aligned on going with the trace route, closing this for now

**Pull request closed without merging**